students/patchcarrier/Lesson06/test_mailroom4.py

```diff
@@ -0,0 +1,187 @@
+import mailroom4 as mr
+import sys
+import io
+import random
+import subprocess
+import glob
+import os
+
+def donor_entry_help(name, amount):
+    """Add the donor with the specified name and donation amount to the global
+    dictionary using the mr.enter_donor() function."""
+    
+    # The enter_donor() function expects two inputs from stdin
+    cmds = f"{name}  {amount}".replace("  ", "\n")
+    sys.stdin = io.StringIO(cmds)
+    mr.enter_donor()
+
+
+def test_donor_entry_1():
+    "Test adding a donation to an existing donor."
+    
+    # loop through each existing donor and add a new random ammount
+    for donor in mr.donors:
+        n_donations = len(mr.donors[donor])
+        donation_amount = random.uniform(0,100)
+        donor_entry_help(donor,donation_amount)
+        
+        # Check that the previous donations are still recorded in the global 
+        # dictionary, and that the new donation got recorded
+        assert len(mr.donors[donor]) == n_donations + 1
+        # Check that the correct donation ammount got added to the global dictionary
+        assert mr.donors[donor][-1] == donation_amount
+    
+    
+def test_donor_entry_2():
+    "Test adding a donation amount for a new donor."
+    
+    new_names = ("Patrick Carrier", "Jim Harbaugh", "Kelly Johnson")
+    
+    for name in new_names:
+        donation_amount = random.uniform(0,100)
+        donor_entry_help(name, donation_amount)
+        
+        # Ensure that each new donor only has one donation recorded and 
+        # that the correct value got recorded in the global dictionary
+        assert len(mr.donors.get(name)) == 1
+        assert mr.donors.get(name)[-1] == donation_amount
+    
+
+def test_donor_entry_3():
+    "Test non-numerical input for donation amount being handled."
+    
+    try:
+        # Attempt to add the non-numerical donation amount 'money'
+        donor_entry_help("Pablo Picasso", 'money')
+    except EOFError:
+        # mailroom.py should continue prompting the user for a numerical input
+        # so the above function call should raise an EOFError since we only
+        # provide the one value: 'money'
+        pass
+
+
+def test_list_donors():
+    "Verify that all of the donor names are printed to stdout"
+    
+    # Re-direct output to the terminal to the StringIO object output
+    output = io.StringIO()
+    sys.stdout = output
+    mr.list_donors()
+    # Reset stdout
+    sys.stdout = sys.__stdout__
```

Comment:
the idea you have is right, but better to use built in offerings:
https://docs.pytest.org/en/latest/capture.html#accessing-captured-output-from-a-test-function